### PR TITLE
server: Remove authentication-related session creation

### LINKED ISSUE #130: Get client IP address while client connects to smtp server?
Dear @emersion

First of all, thanks for the great work. 👍 

We're trying to implement an internal SMTP server, and we'd like to reject client as earlier as possible if client IP address is not whitelisted. Currently it's ok to get client IP in `Mail()` method of [`Session` interface](https://pkg.go.dev/github.com/emersion/go-smtp#Session), but we'd like to reject it even earlier. For example, when the client connects to the smtp server.

Could you share some idea of how we can achieve this? Thank you. :)

### LINKED ISSUE #136: Add a way to hook into the HELO commands
Similar to #130, a hook to the HELO commands can be useful to implement security depending on the value of the argument provided.

Examples: check the domain syntax, its existence, whether it's correctly configured to accept emails, etc.

### PR DESCRIPTION
**Pull request by foxcpp**
Closes #136.
Closes #130.

Here is the backend API change that addresses my concerns from #114 and also resolves two other issues.

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -90,15 +90,13 @@ func NewServer(be Backend) *Server {
...
+					sess := conn.Session()
+					if sess == nil {
+						panic("No session when AUTH is called")
 					}
 
-					conn.SetSession(session)
-					return nil
-				})
+					return sess.AuthPlain(username, password)
+					})
```
Comment (emersion) [inline server.go:99]: Indentation is off here

File: conn.go [hunk 2]
```diff
@@ -552,7 +541,7 @@ func (c *Conn) handleAuth(arg string) {
 				c.WriteResponse(smtpErr.Code, smtpErr.EnhancedCode, smtpErr.Message)
 				return
 			}
-			c.WriteResponse(454, EnhancedCode{4, 7, 0}, err.Error())
+			c.WriteResponse(502, EnhancedCode{5, 7, 0}, err.Error())
```
Comment (emersion) [inline conn.go:544]: Hm, is this code more appropriate somehow?

Comment (foxcpp) [reply conn.go]: Right, on a second thought I think 454 is better. Reverted the change.

File: server.go (hunk 1 above)
Comment (foxcpp) [reply server.go]: [Fixed](https://github.com/emersion/go-smtp/compare/ec3cfb6648fb2d78cdc8907c75a60970e7f69333..03f13880e866f1351daca7d62d22936ea3d1a6c7)

File: conn.go (hunk 2 above)
Comment (emersion) [reply conn.go]: Note, I haven't actually looked at the RFC, just wanted to know the motivation for the change.

**Review by emersion — APPROVED** [review]
LGTM, thanks!

File: conn.go (hunk 2 above)
Comment (foxcpp) [reply conn.go]: 502 is completely unrelated to authentication whereas 454 is "temporary authentication failure". Not sure where I got 502 from.

**Comment by kayrus** [comment]
https://github.com/emersion/go-smtp-proxy must be updated as well

**Pull request merged**